waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,483 @@
+
+(def service-id->failed-instances-transient-store (atom {}))
+(def instance-x->failed-instances-transient-store (atom {}))
+(def k8s-api-auth-str (atom nil))
+
+;; XXX - this shouldn't be a global constant (it should be read from the config file)
+(def k8s-name-length-limit 63)
+
+(def k8s-timestamp-format (DateTimeFormat/forPattern "yyyy-MM-dd'T'HH:mm:ss'Z'"))
+
+(defn- timestamp-str->datetime [k8s-timestamp-str]
+  (utils/str-to-date k8s-timestamp-str k8s-timestamp-format))
+
+(defn- service-id->app-name [service-id]
+  (let [[_ app-prefix x y z] (re-find #"([^-]+)-(\w{8})(\w+)(\w{8})$" service-id)
+        app-name-max-length (- k8s-name-length-limit 6)
+        suffix (if (< app-name-max-length 48)
+                 (str \- x z)
+                 (str \- x y z))
+        prefix-max-length (- app-name-max-length (count suffix))
+        app-prefix' (cond-> app-prefix
+                      (< prefix-max-length (count app-prefix))
+                      (subs 0 prefix-max-length))]
+      (str app-prefix' suffix)))
+
+(defn- kw->str
+  "Convert a keyword to a string, preserving the keyword's namespace (if any)."
+  [^clojure.lang.Keyword kw]
+  (str (.-sym kw)))
+
+(defn- as-json
+  "Convert Clojure data structures to JSON, preserving namespaces on keys."
+  [data]
+  (json/write-str data :key-fn #(if (keyword? %) (kw->str %) %)))
+
+(defn- replicaset->Service
+  ;; FIXME - catch exceptions and return nil
+  [replicaset]
+  (let [replicaset-status (:status replicaset)
+        ready (:readyReplicas replicaset-status 0)
+        requested (get-in replicaset [:spec :replicas])
+        started (:replicas replicaset-status 0)
+        available (:availableReplicas replicaset-status 0)
+        unavailable (:unavailableReplicas replicaset-status 0)]
+    (log/info "ReplicaSet info" replicaset)
+    (scheduler/make-Service
+      {:app-name (get-in replicaset [:metadata :name])
+       :id (get-in replicaset [:metadata :annotations :waiter/service-id])
+       :instances ready
+       :namespace (get-in replicaset [:metadata :namespace])
+       :task-count requested
+       :task-stats {:running ready
+                    :staged (- requested ready)
+                    :healthy available
+                    :unhealthy unavailable}})))
+
+(defn- pod->instance-id
+  ([pod] (pod->instance-id pod (get-in pod [:status :containerStatuses 0 :restartCount])))
+  ([pod restart-count]
+   (let [pod-name (get-in pod [:metadata :name])
+         instance-suffix (subs pod-name (- (count pod-name) 5))
+         service-id (get-in pod [:metadata :annotations :waiter/service-id])]
+     (str service-id \. instance-suffix \- restart-count))))
+
+(defn- track-failed-instances [live-instance pod]
+  (when-let [newest-failure (get-in pod [:status :containerStatuses 0 :lastState :terminated])]
+    (let [service-id (:service-id live-instance)
+          instance-id (:id live-instance)
+          instance-x (pod->instance-id pod \X)
```

Comment:
What is `instance-x`? Maybe a better name?

Reply:
Removed while refactoring.